### PR TITLE
Add SMBUS driver API, Intel PCH SMBUS driver, sample and tests

### LINKED ISSUE #37101: RFC: API change: Add SMbus support to I2C API
## Introduction

Extend I2C driver with SMbus support. 
This proposal is greatly inspired from the [I2C/SMBus subsystem from Linux](https://www.kernel.org/doc/html/latest/i2c/index.html)

### Problem description

SMbus block read cannot be performed with the current API. `i2c_api_full_io_t` can be used to perform SMbus operations, however SMbus block read is impossible because of the `Byte count` field.

### Proposed change

Add SMbus API by extending the existing I2C API.

1. There are two main differences between I2C and SBbus:
- SMbus has a speed limit of minimum 10KHz and maximum of 100KHz. I suggest to not implement any limitation in this regard, let the user be in charge of this.
- SMbus operations have a timeout of 35ms. Each SMbus operation must return ETIMEDOUT after this amount of time

2. Not all hardware can support all SMbus operations. A new function should be added for the user to check the driver functionality. 

3. New functions should be added for all SMbus operations: quick command, read/write byte/word, read/write block data

### Proposed change (Detailed)

Extend `i2c_driver_api` with the following functionality:
```
int i2c_api_smbus_scan_t(uint16_t address) - Checks if the device with a given address is present on bus
uint32_t i2c_api_get_functionality(const struct device *dev) - returns bit encoded functionality of the driver. See defines below

// SMbus specific commands
int i2c_api_smbus_quick_command_t(const struct device *dev)
int i2c_api_smbus_read_byte_t(const struct device *dev)
int i2c_api_smbus_write_byte_t(const struct device *dev, uint8_t value)
int i2c_api_smbus_read_byte_data_t(const struct device *dev, uint8_t command)
int i2c_api_smbus_write_byte_data_t(const struct device *dev, uint8_t command, uint8_t value)
int i2c_api_smbus_read_word_data_t(const struct device *dev, uint8_t command)
int i2c_api_smbus_write_word_data_t(const struct device *dev, uint8_t command, uint16_t value)
int i2c_api_smbus_read_block_data_t(const struct device *dev, uint8_t command, uint8_t *data, size_t data_len)
int i2c_api_smbus_write_block_data_t(const struct device *dev, uint8_t command, uint8_t data. size_t len)
```

Add the following defines to describe driver functionality:
```
I2C_FUNC_I2C - Plain i2c-level commands (Pure SMBus adapters typically can not do these)
I2C_FUNC_10BIT_ADDR - Handles the 10-bit address extensions
I2C_FUNC_SMBUS_QUICK - Handles the SMBus write_quick command
I2C_FUNC_SMBUS_READ_BYTE - Handles the SMBus read_byte command
I2C_FUNC_SMBUS_WRITE_BYTE - Handles the SMBus write_byte command
I2C_FUNC_SMBUS_READ_BYTE_DATA - Handles the SMBus read_byte_data command
I2C_FUNC_SMBUS_WRITE_BYTE_DATA - Handles the SMBus write_byte_data command
I2C_FUNC_SMBUS_READ_WORD_DATA - Handles the SMBus read_word_data command
I2C_FUNC_SMBUS_WRITE_WORD_DATA - Handles the SMBus write_byte_data command
I2C_FUNC_SMBUS_PROC_CALL - Handles the SMBus process_call command
I2C_FUNC_SMBUS_READ_BLOCK_DATA - Handles the SMBus read_block_data command
I2C_FUNC_SMBUS_WRITE_BLOCK_DATA - Handles the SMBus write_block_data command
```

Each of i2c_api_smbus_* should return ETIMEDOUT when the timeout duration has passed.

### Dependencies

All I2C drivers. 
At least `i2c_api_get_functionality` should be implemented in each driver.

## Alternatives

Add a new flag for `i2c_msg` to instruct `i2c_api_full_io_t` to read the no. of bytes received at the first byte read.

### PR DESCRIPTION
**Pull request by finikorg**
On Architecture WG meeting we agreed to have SMBus API in their own tree to simplify I2C and SMBus.

PR includes following components after request from @galak:
- SMBus subsystem API
- Intel PCH SMBus driver
- Qemu PCH driver tests
- Native Posix emulation tests
- SMBus sample with SMBus shell support

Can be verified with qemu_x86_64:
```
$ west build -b qemu_x86_64 zephyr/samples/drivers/smbus/ -p -t run
...
*** Booting Zephyr OS build zephyr-v3.3.0-1231-g00a4a871b5ba ***
Start SMBUS shell sample qemu_x86_64
uart:~$ smbus 
smbus - smbus commands
Subcommands:
  quick            :SMBus Quick command
                    Usage: quick <device> <addr>
  scan             :Scan SMBus peripheral devices command
                    Usage: scan <device>
  byte_read        :SMBus: byte read command
                    Usage: byte_read <device> <addr>
  byte_write       :SMBus: byte write command
                    Usage: byte_write <device> <addr> <value>
  byte_data_read   :SMBus: byte data read command
                    Usage: byte_data_read <device> <addr> <cmd>
  byte_data_write  :SMBus: byte data write command
                    Usage: byte_data_write <device> <addr> <cmd> <value>
  word_data_read   :SMBus: word data read command
                    Usage: word_data_read <device> <addr> <cmd>
  word_data_write  :SMBus: word data write command
                    Usage: word_data_write <device> <addr> <cmd> <value>
  block_write      :SMBus: Block Write command
                    Usage: block_write <device> <addr> <cmd> [<byte1>, ...]
  block_read       :SMBus: Block Read command
                    Usage: block_read <device> <addr> <cmd>
...
uart:~$ smbus scan smbus0 
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:             -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: 50 51 52 53 54 55 56 57 -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --                         
8 devices found on smbus0
...
```

Testing can be done with following command:

```
zephyrproject$ twister -c -t smbus -v --all
...
INFO    - 1126968/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.debug PASSED (build)
INFO    - 1126969/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.debug PASSED (build)
INFO    - 1126970/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api.no_smbalert PASSED (qemu 2.384s)
INFO    - 1126971/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api.no_host_notify PASSED (qemu 2.458s)
INFO    - 1126972/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.no_host_notify PASSED (build)
INFO    - 1126973/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.no_host_notify PASSED (build)
INFO    - 1126974/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.no_smbalert PASSED (build)
INFO    - 1126975/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.no_smbalert PASSED (build)
INFO    - 1126976/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api.debug PASSED (qemu 5.045s)
INFO    - 1126977/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api_stats PASSED (build)
INFO    - 1126978/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api_stats PASSED (qemu 3.604s)
INFO    - 1126979/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api_stats PASSED (build)
INFO    - 1126980/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.access.mmio PASSED (build)
INFO    - 1126981/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.access.mmio PASSED (build)
INFO    - 1126982/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api    PASSED (qemu 2.284s)
INFO    - 1126983/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api    PASSED (build)
INFO    - 1126984/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api    PASSED (build)
INFO    - 1126985/1126992 native_posix              tests/drivers/smbus/smbus_emul/drivers.smbus.emul.peripheral.interrupt PASSED (native 0.139s)
INFO    - 1126986/1126992 native_posix              tests/drivers/smbus/smbus_emul/drivers.smbus.emul  PASSED (native 0.090s)
INFO    - 1126987/1126992 ehl_crb                   samples/drivers/smbus/sample.drivers.smbus.shell   PASSED (build)
INFO    - 1126988/1126992 rpl_crb                   samples/drivers/smbus/sample.drivers.smbus.shell   PASSED (build)
INFO    - 1126989/1126992 qemu_x86_64               samples/drivers/smbus/sample.drivers.smbus.shell   PASSED (qemu 2.148s)
INFO    - 1126990/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api_user PASSED (build)
INFO    - 1126991/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api_user PASSED (build)
INFO    - 1126992/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api_user PASSED (qemu 2.283s)

INFO    - 2242 test scenarios (1126992 test instances) selected, 1126967 configurations skipped (1126967 by static filter, 0 at runtime).
INFO    - 25 of 1126992 test configurations passed (100.00%), 0 failed, 0 errored, 1126967 skipped with 0 warnings in 273.92 seconds
INFO    - In total 101 test cases were executed, 4539671 skipped on 552 out of total 552 platforms (100.00%)
INFO    - 9 test configurations executed on platforms, 16 test configurations were only built.
...
```

Fixes #37101

### CODE REVIEW

**Comment by stephanosio** [comment]
Rebased to fix CI failure

**Comment by teburd** [comment]
I wonder if the route might be to treat an smbus controller as a super set or specialization of i2c instead, much like the i3c API is doing.

Please take a look at how that is being done in #46087

**Comment by finikorg** [comment]
> I wonder if a better route might be to treat an smbus controller as a super set or specialization of i2c instead, much like the i3c API is doing.
> 
> Please take a look at how that is being done in #46087

I wanted to make things simple. I think we only need those smbus functions to be added.

**Comment by finikorg** [comment]
> Architecture WG meeting:
> 
> * @henrikbrixandersen and @nashif argue that this should be its own API
> * @andyross mentions that the quirks in both smbus and I2c would likely make it impractical to share code
> * @finikorg agrees to move the API out of `i2c.h` into its own

I pushed updated version of the API.

**Review by galak — DISMISSED** [review]
Don't we need at least on driver implementation and sample?

**Comment by finikorg** [comment]
> Don't we need at least on driver implementation and sample?

Hi @galak, I wanted to get feedback for the API. I will add also driver implementation.

**Comment by finikorg** [comment]
> Don't we need at least on driver implementation and sample?

Added Intel PCH driver and sample, can be verified with qemu_x86_64

**Review by jhedberg — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

File: drivers/smbus/intel_pch_smbus.c [hunk 1]
```diff
@@ -53,6 +54,21 @@ struct pch_data {
...
+	struct k_work smb_alert_work;
+
+	/* Host Notify callback list */
+	sys_slist_t host_notify_cbs;
+	/* Host Notify work */
+	struct k_work host_notify_work;
+	/* Host Notify peripheral device address */
+	uint8_t notify_addr;
+	/* Host Notify data received */
+	uint16_t notify_data;
```
Comment (MaureenHelm) [inline drivers/smbus/intel_pch_smbus.c:68]: Wrap in `ifdef CONFIG_SMBUS_INTEL_PCH_HOST_NOTIFY`?

**Review by MaureenHelm — DISMISSED** [review]
Nice work

File: drivers/smbus/intel_pch_smbus.c (hunk 1 above)
Comment (finikorg) [reply drivers/smbus/intel_pch_smbus.c]: Added wrap and tests

**Review by MaureenHelm — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by teburd — APPROVED** [review]
really clean and thorough pr

**Pull request merged**